sys/vfs_readdir.c

```diff
@@ -0,0 +1,41 @@
+#include <common.h>
+#include <dirent.h>
+#include <malloc.h>
+#include <uio.h>
+#include <vfs.h>
+
+int readdir_generic(vnode_t *v, uio_t *uio, readdir_ops_t *ops) {
+  void *it = ops->first(v);
+  dirent_t *dir;
+  off_t offset = 0;
+  int error;
+
+  /* Locate proper directory based on offset */
+  for (; it; it = ops->next(it)) {
+    unsigned reclen = _DIRENT_RECLEN(dir, ops->namlen_of(it));
+    if (offset + reclen > uio->uio_offset) {
+      assert(it == NULL || offset == uio->uio_offset);
+      break;
+    }
+    offset += reclen;
+  }
+
+  for (; it; it = ops->next(it)) {
+    unsigned namlen = ops->namlen_of(it);
+    unsigned reclen = _DIRENT_RECLEN(dir, namlen);
+
+    if (uio->uio_resid < reclen)
+      break;
+
+    dir = kmalloc(M_TEMP, reclen, M_ZERO);
```

Comment:
[FreeBSD allocates dirent_t on stack](http://bxr.su/FreeBSD/sys/fs/tmpfs/tmpfs_subr.c#1193), but now there is no point in doing this since we have zero length array at the end of dirent.